lib/runners/workspace.rb

```diff
@@ -1,85 +1,82 @@
-    attr_reader :base_dir
-    attr_reader :head_dir
-    attr_reader :git_ssh_path
+    attr_reader :options, :working_dir, :trace_writer, :retryable_sleep
 
-    def initialize(working_dir:, head_dir:, base_dir:, git_ssh_path:)
+    # @param options [Runners::Options]
+    # @param working_dir [Pathname]
+    # @param trace_writer [Runners::TraceWriter]
+    def initialize(options:, working_dir:, trace_writer:)
+      @options = options
       @working_dir = working_dir
-      @base_dir = base_dir
-      @head_dir = head_dir
-      @git_ssh_path = git_ssh_path
-    end
-
-    def calculate_changes
-      Changes.calculate(base_dir: base_dir, head_dir: head_dir, working_dir: working_dir)
+      @trace_writer = trace_writer
+      @retryable_sleep = -> (n) { n + 1 }
     end
 
-    def self.open(base:, base_key:, head:, head_key:, ssh_key:, working_dir:, trace_writer:)
-      prepare_ssh(ssh_key, trace_writer: trace_writer) do |ssh_path|
+    # @yieldparam git_ssh_path [Pathname, nil]
+    # @yieldparam changes [Runners::Changes]
+    def open
+      prepare_ssh do |git_ssh_path|
         Dir.mktmpdir do |base_dir|
           base_path = Pathname(base_dir)
           Dir.mktmpdir do |head_dir|
             head_path = Pathname(head_dir)
 
             trace_writer.header "Setting up source code"
 
-            prepare_in_dir(:base, base, base_key, base_path, trace_writer: trace_writer) if base
-            prepare_in_dir(:head, head, head_key, head_path, trace_writer: trace_writer)
+            if base = options.base
+              prepare_in_dir(:base, base, options.base_key, base_path)
+            end
+            prepare_in_dir(:head, options.head, options.head_key, head_path)
 
             trace_writer.message "Copying head to working dir..." do
               FileUtils.copy_entry(head_path, working_dir)
             end
 
-            yield new(working_dir: working_dir, base_dir: base_path, head_dir: head_path, git_ssh_path: ssh_path)
+            trace_writer.header "Calculating changes between head and base"
+            trace_writer.message "Running..." do
+              yield git_ssh_path, Changes.calculate(base_dir: base_path, head_dir: head_path, working_dir: working_dir)
+            end
           end
         end
       end
     end
 
-    def self.prepare_ssh(ssh_key, trace_writer:)
-      trace_writer.header "Setting up SSH config"
+    private
 
-      if ssh_key
-        Dir.mktmpdir do |ssh_dir|
+    def prepare_ssh
+      if ssh_key = options.ssh_key
+        Dir.mktmpdir do |dir|
           trace_writer.message "Preparing SSH config..."
 
-          config_path = ssh_dir + "/config"
-          key_path = ssh_dir + "/key"
-          git_ssh_path = ssh_dir + "/run.sh"
+          config_path = Pathname(dir) / 'config'
+          key_path = Pathname(dir) / 'key'
+          script_path = Pathname(dir) / 'run.sh'
 
-          File.write(config_path, <<~SSH_CONFIG)
+          config_path.write(<<~SSH_CONFIG)
```

Comment:
[question] Is `perm: 0600` necessary also here?

```suggestion
          config_path.write(<<~SSH_CONFIG, perm: 0600)
```

Reply:
I don't know why it's been working, but to set permission is required according to ssh_config(5). I'll set it.